### PR TITLE
Avoid invoking the hir_crate query to traverse the HIR

### PR DESCRIPTION
**Pull request by cjgillot**
Walking the HIR tree is done using the `hir_crate` query. However, this is unnecessary, since `hir_owner(CRATE_DEF_ID)` provides the same information. Since depending on `hir_crate` forces dependents to always be executed, this leads to unnecessary work.

By splitting HIR and attributes visits, we can avoid an edge to `hir_crate` when trying to visit the HIR tree.

### CODE REVIEW

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 1]
```diff
@@ -519,6 +519,22 @@ impl<'hir> Map<'hir> {
         }
     }
 
+    /// Walks the contents of a crate. See also `Crate::visit_all_items`.
+    pub fn walk_crate(self, visitor: &mut impl Visitor<'hir>) {
```
Comment (Aaron1011) [inline compiler/rustc_middle/src/hir/map/mod.rs:523]: Could you rename this to something like `walk_crate_without_crate_attrs`, to make it clear that it doesn't walk everything in the crate?

**Comment by pnkfelix** [comment]
A performance run [was done](https://github.com/rust-lang/rust/pull/88435#issuecomment-907700333) on this PR, but for some reason categorization failed.

The aforementioned performance run predicted that this would be a small win, at most.

However, during weekly performance triage, it was tagged as having [mixed results on rustc instruction counts](https://perf.rust-lang.org/compare.html?start=e30b68353fe22b00f40d021e7914eeb78473b3c1&end=7849e3e9dda60e8ec826ee245c6b180e73911b37&stat=instructions:u).

Tagging as perf-regression mostly as hint that maybe someone should go and figure out why there was a discrepancy between what was predicted during PR's run versus what actually happened when it landed on nightly.

**Comment by Mark-Simulacrum** [comment]
Trying to reproduce regressions locally in cachegrind led to a few *improvements* -- and the new significance algorithm puts them close to the significance threshold. So that seems consistent with the original results, and the new ones are probably closer to noise than actual meaningful changes. Marking as triaged; doesn't seem important to investigate further.

**Pull request merged**